File: src/test/scala/com/exasol/cloudetl/transform/DefaultTransformationTest.scala
package com.exasol.cloudetl.transform

import com.exasol.cloudetl.constants.Constants.TRUNCATE_STRING
import com.exasol.cloudetl.helper.StringGenerator.getRandomString
import com.exasol.cloudetl.storage.StorageProperties

import org.scalatest.funsuite.AnyFunSuite

class DefaultTransformationTest extends AnyFunSuite {

  val longString = getRandomString(2000010)
  val values = Array[Any](1, 3.14, "abc", longString).map(_.asInstanceOf[Object])

  test("truncates strings larger than maximum varchar size") {
    val properties = StorageProperties(Map(TRUNCATE_STRING -> "true"))
    val expected = Array[Any](1, 3.14, "abc", longString.substring(0, 2000000)).map(_.asInstanceOf[Object])
    assert(new DefaultTransformation(properties).transform(Array.from(values)) === expected)
  }

  test("passes if there is no strings larger than maximum varchar size") {
    val properties = StorageProperties(Map(TRUNCATE_STRING -> "true"))
    val input = Array("abc", getRandomString(255), getRandomString(2000000)).map(_.asInstanceOf[Object])
    assert(new DefaultTransformation(properties).transform(input) === input)
  }

  test("throws strings larger than maximum varchar size if no truncate string parameter") {
    val properties = StorageProperties(Map.empty[String, String])
    val thrown = intercept[IllegalStateException] {
      new DefaultTransformation(properties).transform(Array.from(values))
    }
    assert(thrown.getMessage().startsWith("E-CSE-29"))
  }

  test("throws strings larger than maximum varchar size if truncate string parameter is false") {
    val properties = StorageProperties(Map(TRUNCATE_STRING -> "false"))
    val thrown = intercept[IllegalStateException] {
      new DefaultTransformation(properties).transform(Array.from(values))
    }
    assert(thrown.getMessage().startsWith("E-CSE-29"))
  }

}

File: src/test/java/com/exasol/cloudetl/transform/DefaultTransformationTest.java
package com.exasol.cloudetl.transform;

import static com.exasol.cloudetl.constants.Constants.TRUNCATE_STRING;
import static com.exasol.cloudetl.helper.StringGenerator.getRandomString;
import static org.junit.jupiter.api.Assertions.*;

import java.util.Map;

import org.junit.jupiter.api.Test;

import com.exasol.cloudetl.storage.StorageProperties;

class DefaultTransformationTest {
    private final String longString = getRandomString(2_000_010);
    private final Object[] values = { 1, 3.14, "abc", this.longString };

    @Test
    void truncatesStringsLargerThanMaximumVarcharSize() {
        final StorageProperties properties = new StorageProperties(Map.of(TRUNCATE_STRING, "true"));
        final Object[] expected = { 1, 3.14, "abc", this.longString.substring(0, 2_000_000) };
        assertArrayEquals(expected, new DefaultTransformation(properties).transform(this.values.clone()));
    }

    @Test
    void passesIfThereIsNoStringsLargerThanMaximumVarcharSize() {
        final StorageProperties properties = new StorageProperties(Map.of(TRUNCATE_STRING, "true"));
        final Object[] input = { "abc", getRandomString(255), getRandomString(2_000_000) };
        assertArrayEquals(input, new DefaultTransformation(properties).transform(input.clone()));
    }

    @Test
    void throwsStringsLargerThanMaximumVarcharSizeIfNoTruncateStringParameter() {
        final StorageProperties properties = new StorageProperties(Map.of());
        final IllegalStateException thrown = assertThrows(IllegalStateException.class,
                () -> new DefaultTransformation(properties).transform(this.values.clone()));
        assertTrue(thrown.getMessage().startsWith("E-CSE-29"));
    }

    @Test
    void throwsStringsLargerThanMaximumVarcharSizeIfTruncateStringParameterIsFalse() {
        final StorageProperties properties = new StorageProperties(Map.of(TRUNCATE_STRING, "false"));
        final IllegalStateException thrown = assertThrows(IllegalStateException.class,
                () -> new DefaultTransformation(properties).transform(this.values.clone()));
        assertTrue(thrown.getMessage().startsWith("E-CSE-29"));
    }
}
